narwhal/primary/src/primary.rs

```diff
@@ -550,6 +541,221 @@ impl PrimaryReceiverHandler {
+                    "Unable to find authority with network key {peer_network_key:?}"
+                ))
+            })?;
+        ensure!(
+            header.author == *peer_authority,
+            DagError::NetworkError(format!(
+                "Header author {:?} must match requesting peer {peer_authority:?}",
+                header.author
+            ))
+        );
+
+        debug!(
+            "Processing vote request for {:?} round:{:?}",
+            header, header.round
+        );
+
+        // Clone the round updates channel so we can get update notifications specific to
+        // this RPC handler.
+        let mut rx_consensus_round_updates = self.rx_consensus_round_updates.clone();
+        let mut consensus_round = *rx_consensus_round_updates.borrow();
+        ensure!(
+            consensus_round <= header.round,
+            DagError::TooOld(header.digest().into(), header.round, consensus_round)
+        );
+
+        // If requester has provided us with parent certificates, process them all
+        // before proceeding. This may advance our round, so do it before checking round.
+        let mut notifies = Vec::new();
+        for certificate in request.body().parents.clone() {
+            let (tx_notify, rx_notify) = oneshot::channel();
+            notifies.push(rx_notify);
+            self.tx_certificates
+                .send((certificate, Some(tx_notify)))
+                .await
+                .map_err(|_| DagError::ChannelFull)?;
+        }
+        let mut wait_notifies = futures::future::try_join_all(notifies);
+        loop {
+            tokio::select! {
+                results = &mut wait_notifies => {
+                    let results: Result<Vec<_>, _> = results
+                        .map_err(|e| DagError::ClosedChannel(format!("{e:?}")))?
+                        .into_iter()
+                        .collect();
+                    results?;
+                    break
+                },
+                result = rx_consensus_round_updates.changed() => {
+                    result.unwrap();
+                    consensus_round = *rx_consensus_round_updates.borrow();
+                    ensure!(
+                        consensus_round <= header.round,
+                        DagError::TooOld(header.digest().into(), header.round, consensus_round)
+                    )
+                },
+            }
+        }
+
+        // Ensure we have the parents. If any are missing, the requester should provide them on retry.
+        let (parents, missing) = self.synchronizer.get_parents(header)?;
+        if !missing.is_empty() {
+            return Ok(RequestVoteResponse {
+                vote: None,
+                missing,
+            });
+        }
+
+        // Now that we've got all the required certificates, ensure we're voting on a
+        // current Header.
+        // DO NOT MERGE: There's no reason for us to vote on a Header with old round, is there?
```

Comment:
I plan to allow voting for headers that are newer than our current round as long as all other checks pass, just so that we don't block other certificates from forming just because some internal processes on our own primary happen to be gunked up. Do you foresee any problems with that? (Can still refuse to vote on old headers of course.)

Reply:
@aschran @gdanezis please also see my above comment https://github.com/MystenLabs/sui/pull/5990/files#r1020206612

Reply:
If all checks pass, namely: we see a header at R' > R our current round, and we process all its parent headers which must be at R'-1, then we should be entering R' ourselves. Is that correct? It kind of should be.

Reply:
Yes, the only reason I don't enforce this here is that the proposer might still be processing the new certificates and not have sent us a narwhal round update yet. But that doesn't seem worth blocking a vote on